### DISCUSSION TITLE
How do we adjust the choking algorithm?

### DISCUSSION THREAD

**Discussion by sandeshbhatjr** [anacrolix/torrent#881] [General]
I use the bit torrent protocol to distribute data amongst different peers, all of whom I control. Is there any easy way to completely modify/disable the choking algorithm?

**Comment by anacrolix** [comment]
Yes, I believe there's an aggressive upload option or something, I'll follow up when I'm on a PC

**Reply by anacrolix** [reply]
It looks like if you seed ClientConfig.Seed to true, the client will upload and unchoke everyone without restriction. See Torrent.seeding, and PeerConn.upload.

**Reply by sandeshbhatjr** [reply]
@anacrolix I use `Seed` set to `true` and `DisableAggressiveUpload` to `false`. Looking at the code, it would seem that it should upload. However, it only uploads once the download of all pieces is complete.

I use the following setup: There are three peers: A, B and C. Each peer is a simple client created using your library with `Seed` set to `true` and `DisableAggressiveUpload` set to `false`. Peer C is provided Peer B as a peer, while Peer B is provided Peer A as a peer. That is a peer dependency as follows: Peer A -> Peer B -> Peer C. Peer A has a file of size 10GB. They are all connected via localhost on different ports. Running them together, Peer B first completes its download, and only after it is complete does the Peer C start its download. Isn't it possible for Peer B to start peering for Peer C when it already has some pieces rather than wait until it completes its download.

Am I missing something?

**Reply by sandeshbhatjr** [reply]
Note also that in addition to the above, I also experimented by disabling Peer A halfway throught the download, i.e Peer B has half the pieces and cannot connect to Peer A in this situation. Even then, Peer B does NOT upload to Peer C. It seems like a bug to me.

**Reply by sandeshbhatjr** [reply]
Exploring the code further, the following function seems to decide when to upload or not:
```
func (c *PeerConn) uploadAllowed() bool {
	if c.t.cl.config.NoUpload {
		return false
	}
	if c.t.dataUploadDisallowed {
		return false
	}
	if c.t.seeding() {
		return true
	}
	if !c.peerHasWantedPieces() {
		return false
	}
	// Don't upload more than 100 KiB more than we download.
	if c._stats.BytesWrittenData.Int64() >= c._stats.BytesReadData.Int64()+100<<10 {
		return false
	}
	return true
}
```
Here, upload should be allowed if seeding() is true (assuming NoUpload is set to false, but I'm not so sure about dataUploadDisallowed). Now, when I look at seeding() in `torrent.go`, it is as follows:
```
func (t *Torrent) seeding() bool {
	cl := t.cl
	if t.closed.IsSet() {
		return false
	}
	if t.dataUploadDisallowed {
		return false
	}
	if cl.config.NoUpload {
		return false
	}
	if !cl.config.Seed {
		return false
	}
	if cl.config.DisableAggressiveUpload && t.needData() {
		return false
	}
	return true
}
```
During download, I'm assuming `t.needData()` is true. So when DisableAggressiveUpload is set to false, seeding() should be true (again assuming dataUploadDisallowed is false).

From this, it seems like the bad behaviour is probably coming from somewhere else.

**Reply by anacrolix** [reply]
@sandeshbhatjr thanks for the deep dive. It does seem like you arrived at the same conclusion as me, that uploading should be possible but doesn't appear to happen.

The next place to look is to see if uploads are actually made. I suspect that if you are downloading over localhost, that the outbound connection buffer is completely overloaded with requests for more pieces. You could try looking there, and maybe instrumenting to see if it's the reason. Take a look at the code surrounding https://github.com/anacrolix/torrent/blob/72c1152b395aba402e928601e61fca5ac03792d5/peerconn.go#L306-L313.

**Reply by sandeshbhatjr** [reply]
@anacrolix So, I investigated a bit further, but I am not sure I understand the logic here. My mental model was that the `writeBuffer` is where you store the file data before you upload it. In this case, the `writeBuffer` is not filling up for some reason.

In the loop below, the writer starts off with elements in the `writeBuffer` (which is not empty in the beginning). It then swaps it with `frontBuffer`, and in the subsequent iteration the `writeBuffer` is empty. It remains empty after the first iteration, so basically the loop just waits until the timer runs out, and repeats with an empty `writeBuffer`.

https://github.com/anacrolix/torrent/blob/e10ffd133acfb476b1ebd69f9380a4e300ad75b9/peer-conn-msg-writer.go#L67C1-L91C4

**Reply by sandeshbhatjr** [reply]
On second thought, I think I got the logic wrong above. The upload function just adds to the write buffer via the `sendChunk` fn, right? 

The problem, I think, is the loop over `c.peerRequests` in the upload fn; `c.peerRequests` is always empty. So nothing really happens in the upload fn. I will try to investigate why `peerRequests` is turning out empty. Please do correct me if I am wrong here.

**Reply by anacrolix** [reply]
Yeah I'm pretty sure the buffer swapping stuff works fine, but I could be wrong. If peerRequests is empty that would definitely explain the issues you are seeing.

**Reply by anacrolix** [reply]
I can't see anything that cause it. Do you have an upload rate limit set?

**Reply by sandeshbhatjr** [reply]
@anacrolix Yes, the buffer is fine. It's definitely the peer request being consistently empty. I used `WriteStatus()` method of the `client` to get info about the peers and their connections. 

The peers connect and stay connected, however they do not send any requests for pieces. They send one unchoke message in the beginning and then no messages are sent after that. Maybe you can point me to the code where the logic for requesting pieces is?

I do not use any upload limit in my config.

Thanks again for the support!

**Comment by anacrolix** [comment]
Have you set download priorities on the torrent?

**Reply by sandeshbhatjr** [reply]
I don't think so. My settings are as follows with dataDir, port and remotes passed in as CLI arguments:
```
cfg := NewDefaultClientConfig()
cfg.NoDHT = true
cfg.DisableTrackers = true
cfg.DisablePEX = true
cfg.DisableAggressiveUpload = false
cfg.DataDir = dataDir
cfg.NoDefaultPortForwarding = true
cfg.PieceHashersPerTorrent = 10
cfg.Extensions = peer_protocol.NewPeerExtensionBytes(peer_protocol.ExtensionBitV2, peer_protocol.ExtensionBitFast)
// upload after all pieces are downloaded
cfg.Seed = true
cfg.AlwaysWantConns = true
cfg.AcceptPeerConnections = true
cfg.EstablishedConnsPerTorrent = 20
// upload completed pieces to while remaining pieces are downloading
cfg.DisableAggressiveUpload = false
cfg.ListenPort = int(port)
client, err := torrent.NewClient(&cfg)
```

I then just run it in a loop with a fixed peer list I call remotes (which is []netip.AddrPort) as follows: 
```
t, _ := client.AddTorrentFromFile(torrentFilepath)
if err != nil {
	// do something
}
<-t.GotInfo() // blocking until torrent is added
// handle the case where the file is changed midway?
for {
	select {
	case <-ctx.Done():
		return
	default:
		// monitor status 
		if ((t.Stats().HalfOpenPeers == 0) && (t.Stats().PendingPeers == 0)) {
			for _, remoteNode := range remotes {
				// update peers regularly using peer mapper
				_ = t.AddPeers([]torrent.PeerInfo{
					{
						Source: torrent.PeerSourceDirect,
						Addr: remoteNode,
						Trusted: true,
					},
				})
			}
		}
		if t.BytesMissing() > 0 {
			t.DownloadAll()
		}
		time.Sleep(REFRESH_PEER_INTERVALS) // TODO: This is so hacky!
	}
}
```

**Reply by sandeshbhatjr** [reply]
@anacrolix Any luck figuring out the issue? Or am I doing something wrong with configurations?

**Reply by anacrolix** [reply]
The configuration looks fine.

Just to confirm, you are using the builtin upload/choking implementation, you haven't made any modifications? The issue is that peers won't start uploading until they've fully obtained a torrent.

I think this needs a test of say 10 peers, with a single initial seeder, all connected to each other. The goals would be to confirm that they request different pieces from the seeder and upload to one another throughout. 

I have some tests in place already that do a seeder/leecher/leecher leecher arrangement like the one you described, there may be some where the leecher has a smaller cache than the entire size of the torrent. That should show similar behaviour to what you describe, but might not match exactly.

**Reply by anacrolix** [reply]
@sandeshbhatjr do you mind looking at https://github.com/anacrolix/torrent/discussions/916? I believe it's related. I have added a test there, it should be similar to the problem you were encountering.

**Comment by anacrolix** [comment]
This should be fixed by https://github.com/anacrolix/torrent/commit/528c5f69d657d6edb7359674473eec1ccff10b9a.
